File: modules/focus-noise-generator/android/src/main/java/expo/modules/focusnoisegenerator/FocusNoiseGeneratorModule.kt
package expo.modules.focusnoisegenerator

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.math.PI
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.random.Random

private enum class NoiseType {
  WHITE,
  PINK,
  BROWN;

  companion object {
    fun fromJs(value: String): NoiseType? {
      return entries.firstOrNull { it.name.equals(value, ignoreCase = true) }
    }
  }
}

private class NoiseGeneratorEngine {
  private val sampleRate = 44_100
  private val channelConfig = AudioFormat.CHANNEL_OUT_STEREO
  private val encoding = AudioFormat.ENCODING_PCM_16BIT
  private val minBufferSize = AudioTrack.getMinBufferSize(sampleRate, channelConfig, encoding).coerceAtLeast(4096)

  @Volatile private var currentType = NoiseType.WHITE
  @Volatile private var running = false
  @Volatile private var volume = 0.22f
  @Volatile private var dingSamplesRemaining = 0

  private var audioTrack: AudioTrack? = null
  private var workerThread: Thread? = null
  private var brownState = 0f
  private var pinkA = 0f
  private var pinkB = 0f
  private var pinkC = 0f
  private var dingPhase = 0.0

  fun play(type: NoiseType) {
    currentType = type
    ensureTrack()
    if (running) {
      return
    }

    running = true
    audioTrack?.play()
    workerThread = Thread {
      writeLoop()
    }.also { it.start() }
  }

  fun stop() {
    running = false
    dingSamplesRemaining = 0
    workerThread?.join(250)
    workerThread = null
    audioTrack?.pause()
    audioTrack?.flush()
  }

  fun setVolume(value: Double) {
    volume = value.toFloat().coerceIn(0f, 1f)
  }

  fun playNotice() {
    dingSamplesRemaining = (sampleRate * 0.18).toInt()
    dingPhase = 0.0
  }

  private fun ensureTrack() {
    if (audioTrack != null) {
      return
    }

    audioTrack = AudioTrack(
      AudioAttributes.Builder()
        .setUsage(AudioAttributes.USAGE_MEDIA)
        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
        .build(),
      AudioFormat.Builder()
        .setEncoding(encoding)
        .setSampleRate(sampleRate)
        .setChannelMask(channelConfig)
        .build(),
      minBufferSize,
      AudioTrack.MODE_STREAM,
      AudioManager.AUDIO_SESSION_ID_GENERATE
    ).apply {
      setVolume(1f)
    }
  }

  private fun writeLoop() {
    val frameBuffer = ShortArray(minBufferSize / 2)

    while (running) {
      var index = 0
      while (index < frameBuffer.size) {
        val mixed = (nextNoiseSample() * volume + nextDingSample()).coerceIn(-1f, 1f)
        val pcm = (mixed * Short.MAX_VALUE).toInt().toShort()
        frameBuffer[index++] = pcm
        frameBuffer[index++] = pcm
      }

      audioTrack?.write(frameBuffer, 0, frameBuffer.size)
    }
  }

  private fun nextWhite(): Float {
    return Random.nextFloat() * 2f - 1f
  }

  private fun nextNoiseSample(): Float {
    return when (currentType) {
      NoiseType.WHITE -> nextWhite()
      NoiseType.PINK -> {
        val white = nextWhite()
        pinkA = 0.99765f * pinkA + white * 0.0990460f
        pinkB = 0.96300f * pinkB + white * 0.2965164f
        pinkC = 0.57000f * pinkC + white * 1.0526913f
        (pinkA + pinkB + pinkC + white * 0.1848f) * 0.05f
      }
      NoiseType.BROWN -> {
        val white = nextWhite() * 0.02f
        brownState = max(-1f, min(1f, brownState + white))
        brownState * 3.5f
      }
    }
  }

  private fun nextDingSample(): Float {
    if (dingSamplesRemaining <= 0) {
      return 0f
    }

    val totalSamples = sampleRate * 0.18
    val progress = dingSamplesRemaining / totalSamples
    val envelope = progress.toFloat() * progress.toFloat()
    val sample = sin(dingPhase).toFloat() * envelope * 0.18f
    dingPhase += 2.0 * PI * 880.0 / sampleRate.toDouble()
    dingSamplesRemaining -= 1
    return sample
  }
}

class FocusNoiseGeneratorModule : Module() {
  private val engine = NoiseGeneratorEngine()

  override fun definition() = ModuleDefinition {
    Name("FocusNoiseGenerator")

    AsyncFunction("playNoise") { type: String ->
      val noiseType = NoiseType.fromJs(type)
        ?: throw Exceptions.InvalidArgument("Unsupported noise type: $type")
      engine.play(noiseType)
    }

    AsyncFunction("stopNoise") {
      engine.stop()
    }

    AsyncFunction("setVolume") { value: Double ->
      engine.setVolume(value)
    }

    AsyncFunction("playNotice") {
      engine.playNotice()
    }

    OnDestroy {
      engine.stop()
    }
  }
}

File: modules/focus-noise-generator/ios/FocusNoiseGeneratorModule.swift
import AVFoundation
import ExpoModulesCore

private enum NoiseType: String {
  case white
  case pink
  case brown
}

private final class NoiseEngine {
  private let engine = AVAudioEngine()
  private var sourceNode: AVAudioSourceNode?
  private let sampleRate: Double = 44_100
  private let channelCount: AVAudioChannelCount = 2

  private var currentType: NoiseType = .white
  private var volume: Float = 0.22
  private var brownState: Float = 0
  private var pinkA: Float = 0
  private var pinkB: Float = 0
  private var pinkC: Float = 0
  private var dingFramesRemaining: Int = 0
  private var dingPhase: Float = 0

  init() {
    configureAudioSession()
    setupEngineIfNeeded()
  }

  func play(type: NoiseType) throws {
    currentType = type
    configureAudioSession()
    setupEngineIfNeeded()

    if !engine.isRunning {
      try engine.start()
    }
  }

  func stop() {
    dingFramesRemaining = 0
    if engine.isRunning {
      engine.pause()
    }
  }

  func setVolume(_ value: Float) {
    volume = max(0, min(value, 1))
  }

  func playNotice() {
    dingFramesRemaining = Int(sampleRate * 0.18)
    dingPhase = 0
  }

  private func configureAudioSession() {
    let session = AVAudioSession.sharedInstance()
    do {
      try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
      try session.setActive(true)
    } catch {
      // The module should remain usable even if the session setup is delayed.
    }
  }

  private func setupEngineIfNeeded() {
    guard sourceNode == nil else { return }

    let format = AVAudioFormat(
      commonFormat: .pcmFormatFloat32,
      sampleRate: sampleRate,
      channels: channelCount,
      interleaved: false
    )!

    let node = AVAudioSourceNode { [weak self] _, _, frameCount, audioBufferList -> OSStatus in
      guard let self else { return noErr }

      let ablPointer = UnsafeMutableAudioBufferListPointer(audioBufferList)
      for frame in 0 ..< Int(frameCount) {
        let noiseSample = self.nextNoiseSample() * self.volume
        let dingSample = self.nextDingSample()
        let mixedSample = max(-1, min(1, noiseSample + dingSample))

        for buffer in ablPointer {
          let pointer = buffer.mData?.assumingMemoryBound(to: Float.self)
          pointer?[frame] = mixedSample
        }
      }

      return noErr
    }

    sourceNode = node
    engine.attach(node)
    engine.connect(node, to: engine.mainMixerNode, format: format)
    engine.prepare()
  }

  private func nextWhite() -> Float {
    Float.random(in: -1 ... 1)
  }

  private func nextNoiseSample() -> Float {
    switch currentType {
    case .white:
      return nextWhite()
    case .pink:
      let white = nextWhite()
      pinkA = 0.99765 * pinkA + white * 0.0990460
      pinkB = 0.96300 * pinkB + white * 0.2965164
      pinkC = 0.57000 * pinkC + white * 1.0526913
      return (pinkA + pinkB + pinkC + white * 0.1848) * 0.05
    case .brown:
      let white = nextWhite() * 0.02
      brownState = max(-1, min(1, brownState + white))
      return brownState * 3.5
    }
  }

  private func nextDingSample() -> Float {
    guard dingFramesRemaining > 0 else { return 0 }

    let progress = Float(dingFramesRemaining) / Float(sampleRate * 0.18)
    let envelope = progress * progress
    let sample = sin(dingPhase) * envelope * 0.18
    dingPhase += Float((2 * Double.pi * 880) / sampleRate)
    dingFramesRemaining -= 1
    return sample
  }
}

public class FocusNoiseGeneratorModule: Module {
  private let noiseEngine = NoiseEngine()

  public func definition() -> ModuleDefinition {
    Name("FocusNoiseGenerator")

    AsyncFunction("playNoise") { (type: String) throws in
      guard let noiseType = NoiseType(rawValue: type) else {
        throw Exception(name: "invalid_type", description: "Unsupported noise type: \(type)")
      }

      try self.noiseEngine.play(type: noiseType)
    }

    AsyncFunction("stopNoise") {
      self.noiseEngine.stop()
    }

    AsyncFunction("setVolume") { (value: Double) in
      self.noiseEngine.setVolume(Float(value))
    }

    AsyncFunction("playNotice") {
      self.noiseEngine.playNotice()
    }
  }
}
